cachito/workers/pkg_managers/pip.py

```diff
@@ -1448,6 +1477,115 @@ def _sdist_preference(sdist_pkg):
     return yanked_pref, filetype_pref
 
 
+def _download_vcs_package(requirement, pip_deps_dir, pip_raw_repo_name, nexus_auth):
+    """
+    Fetch the source for a Python package from VCS (only git is supported).
+
+    After downloading, upload this package to the Nexus raw repository and on subsequent
+    downloads, reuse the uploaded asset instead of fetching from VCS again.
+
+    :param PipRequirement requirement: VCS requirement from a requirements.txt file
+    :param Path pip_deps_dir: The deps/pip directory in a Cachito request bundle
+    :param str pip_raw_repo_name: Name of the Nexus raw repository for Pip
+    :param requests.auth.AuthBase nexus_auth: Authorization for the Nexus raw repo
+
+    :return: Path to downloaded file
+    """
+    git_info = _extract_git_info(requirement.url)
+
+    namespace_parts = git_info["namespace"].split("/")
+    repo_name = git_info["repo"]
+    ref = git_info["ref"]
+
+    # Download to e.g. deps/pip/github.com/namespace/repo
+    package_dir = pip_deps_dir.joinpath(git_info["host"], *namespace_parts, repo_name)
+    package_dir.mkdir(parents=True, exist_ok=True)
+
+    filename = f"{repo_name}-external-gitcommit-{ref}.tar.gz"
+    download_path = package_dir / filename
+
+    # Check if we already have the raw component
+    raw_component_name = f"{repo_name}/{filename}"
+    log.debug("Looking for raw component %r in %r repo", raw_component_name, pip_raw_repo_name)
+    download_url = nexus.get_raw_component_asset_url(pip_raw_repo_name, raw_component_name)
+
+    if download_url is not None:
+        log.debug("Found raw component, will download from %r", download_url)
+        download_binary_file(download_url, download_path, auth=nexus_auth)
+    else:
+        log.debug("Raw component not found, will fetch from git")
+        repo = Git(git_info["url"], ref)
+        repo.fetch_source()
+        log.debug("Fetched package, uploading as %r to %r", raw_component_name, pip_raw_repo_name)
+        upload_raw_package(
+            pip_raw_repo_name,
+            repo.sources_dir.archive_path,
+            dest_dir=repo_name,
+            filename=filename,
+            is_request_repository=False,
+        )
+        # Move downloaded archive to expected download path
+        shutil.move(repo.sources_dir.archive_path, download_path)
```

Comment:
Interesting. So, in this case, we would not be using the current feature to re-use sources avaiable in `fetch_source`?

Reply:
Well, we only want to fetch from git the first time, after that we want to fetch from Nexus right? I can use `copy()` here, though it might be a minor waste of space, WDYT?

Reply:
Nevermind, this is a bad idea, if someone uses X as their app and someone else uses X as a dependency, this could delete the archive that was created for the app. I will use `copy()` instead